Add database servers for use by the instance manager.

smarnach opened the pull request.

smarnach committed:
Add new MySQL server for the instance manager.

smarnach committed:
Add PostgreSQL server.

smarnach committed:
Add MongoDB server for the instance manager.

smarnach committed:
Use hashes instead of version tags in requirements.yml.

smarnach committed:
Configure PostgreSQL server to listen for external connections.

antoviaque commented on host_vars/mongodb.opencraft.hosting/public.yml:
```diff
@@ -0,0 +1,11 @@
+
+TARSNAP_KEY: "{{ MONGODB_TARSNAP_KEY }}"
+SANITY_CHECK_SNITCH: "{{ MONGODB_SANITY_CHECK_SNITCH }}"
+TARSNAP_BACKUP_SNITCH: "{{ MONGODB_TARSNAP_BACKUP_SNITCH }}"
```
Can we keep the public repo free of references to the individual servers, to keep it generic? Is it possible for example to move the `host_vars` directory to the private repository, which is the one with the knowledge of the different hosts? Or, if not, a dict in the private vars containing the values for each host, and just use a sample default in the (public) `group_vars`?

smarnach replied:
The host_vars directory can be in either the same directory as the playbook or the same directory as the inventory, so it should be possible to simply move the whole host_vars directory to the private repo.

@jbzdak Do you have any concerns about this change?

jbzdak replied:
@antoviaque @smarnach This is not the shining point of Ansible. Ansible variable precedence is [defined here](http://docs.ansible.com/ansible/playbooks_variables.html#variable-precedence-where-should-i-put-a-variable), but I here is the relevant part (lower entries have higher precedence: 

- inventory group_vars
- inventory host_vars
- playbook group_vars
- playbook host_vars
- ( ... )
- extra vars (always win precedence)

For now public repositories contain variables that fall into 'playbook'  category, and private variables fall into `inventory category`. So if we started using `host_vars` in secret repository `group_vars` from public repository would have precedence, which is IMO non-intuitive, and could be error prone (as one naturally assumes that `host_vars` override `group_vars`). Private vars have nice property of having biggest priority. .

On the other hand having one big `private-vars` (that is not easily mergeable) is also a pain. 

Cleanest solution would be to move all `host_vars` and `group_vars` to secret repository, but this would make everything less useful to others (which could be fixed by having updated documentation).

smarnach replied:
@jbzdak Thanks, right.  I remembered there was a problem with precedence, but didn't remember the specifics.  Is there some other work-around we could use?  E.g. how about symlinking a directory from the secret repo to host_vars in the public repo?  That would be a terrible hack, but it would result in a natural variable precedence.

jbzdak replied:
@smarnach This would work, another hack that would work (and might be considered somewhat less terrible) would be to link `group_vars` and playbooks from public repo to secret repo.

smarnach replied:
@jbzdak Yes, this sounds better, since we wouldn't have any ugly artifacts in the public repo.  I'll try that and see how it goes.

smarnach replied:
I've implemented the symlink approach, and I think it's much nicer than the previous solution.  If we decide to keep it, we should implement it for the deployment-secrets-services repo as well.

smarnach committed:
Move host_vars directory to private repository.

smarnach committed:
Use branch name instead of hash for geerlingguy.security.

Until the PR is merged, the hash isn't stable and might become invalid when
reabasing.

jbzdak commented:
:+1:
